server: set open file descriptors limits to the hard limit each time

BramGruneir opened the pull request:
RocksDB seems to prefer a value of around 5000 open file descriptors per
instance.  This change sets the soft limit to the hard limit.  If the total
is still lower than the recommended settings it will throw an error.

This change also enables us to pass the max_open_files option to rocksdb.

Part of #7562.
See also https://github.com/cockroachdb/docs/issues/436, https://github.com/cockroachdb/docs/issues/53

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7747)

<!-- Reviewable:end -->

BramGruneir committed:
server: set open file descriptors limits to the hard limit each time

RocksDB seems to prefer a value of around 5000 open file descriptors per
instance.  This change sets the soft limit to the recommend limit or max limit
whichever is smaller.  If the total is still lower than the minimum settings it
will throw an error.

This change also enables us to pass the max_open_files option to rocksdb.

Part of #7562.

BramGruneir commented:
@jseldess, here's that PR.

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQIR7HFUPoNl6E6uCo:-KMQIR7HFUPoNl6E6uCp:1232049318) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L62)):*
> ```Go
> 	defaultReservationsEnabled      = true
> 	minimumOpenFileLimitPerStore    = 256
> 	suggestedOpenFileLimitPerStore  = 5000
> ```

Should this be `= engine.DefaultMaxOpenFiles`?

`s/suggested/desired/g`?

---

*[server/context.go, line 218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJ6XHZ0rA0EiMgagX:-KMQJ6XHZ0rA0EiMgagY:82564233) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L218)):*
> ```Go
> 	minimumOpenFileLimit := uint64(physicalStoreCount * minimumOpenFileLimitPerStore)
> 	suggestedOpenFileLimit := uint64(physicalStoreCount * suggestedOpenFileLimitPerStore)
> 	// TODO(bram): Test this out on windows.
> ```

File an issue.

---

*[server/context.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJCkAiu_c8xEbcS8z:-KMQJCkAiu_c8xEbcS9-:62700749) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L229)):*
> ```Go
> 	// The hard/max open file descriptor limit is too low.
> 	if rLimit.Max < minimumOpenFileLimit {
> 		return 0, fmt.Errorf("hard/max limit for open file descriptors is too low (%d), minimum required %d (%d per on disk store)\n%s",
> ```

The `hard/max` terminology is slightly confusing. Can we just use `max`?

---

*[server/context.go, line 238 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQMxjdxGsSSrcEyTbH:-KMQMxjdxGsSSrcEyTbI:-327226823) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L238)):*
> ```Go
> 	// The soft/current open file descriptor limit is higher than the suggested
> 	// value, we can just use the default value.
> 	if rLimit.Cur > suggestedOpenFileLimit {
> ```

I'm curious why you do these initial checks. Is there a harm to always bumping `rLimit.Cur` to `rLimit.Max`?

---

*[server/context.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQMhoqJKr32tJMGR8c:-KMQMhoqJKr32tJMGR8d:215869909) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L253)):*
> ```Go
> 			return 0, err
> 		}
> 		// Sadly, the soft/current limit does not set to correctly, (e.g. OSX)
> ```

The grammar in this sentence seems off ("does not set to").

---

*[server/context.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQN7Y6yByy-TSc-3FZ:-KMQN7Y6yByy-TSc-3F_:-952224479) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L287)):*
> ```Go
> 		if physicalStoreCount == 1 {
> 			plural = ""
> 		}
> ```

See `util.Pluralize()`.

---

*[server/context.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQNBTj-vrhz812etfY:-KMQNBTj-vrhz812etfZ:-1795605485) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L320)):*
> ```Go
> 	defer cache.Release()
> 
> 	var onDiskStores int
> ```

`s/onDisk/physical/g` to match terminology above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 217 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQPQk7BotYErzCm18N:-KMQPQk7BotYErzCm18O:-584112994) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L217)):*
> ```Go
> func setOpenFileLimit(physicalStoreCount int) (int, error) {
> 	minimumOpenFileLimit := uint64(physicalStoreCount * minimumOpenFileLimitPerStore)
> 	suggestedOpenFileLimit := uint64(physicalStoreCount * suggestedOpenFileLimitPerStore)
> ```

We also need to allow room for network file descriptors.

---

*[server/context.go, line 218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJ6XHZ0rA0EiMgagX:-KMQNhX-R3YTBfHbrRsP:-2080599593) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> File an issue.
</details>
I don't think it's worth filing issues for all the different things that will need work when we start supporting windows. In this case it won't compile because `sycall.[GS]etrlimit` don't exist on windows so we'll need to move this code under a build tag. 

---

*[server/context.go, line 238 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQMxjdxGsSSrcEyTbH:-KMQOfMpXjHtr1X7TTno:-639380162) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L238)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm curious why you do these initial checks. Is there a harm to always bumping `rLimit.Cur` to `rLimit.Max`?
</details>
File descriptors consume some system-wide resources in the kernel. I think that even if the system sets the hard limit to "unlimited" we still want to set some reasonable limit for ourselves so that there's still an upper bound on how many descriptors we can open.

---

*[server/context.go, line 243 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQQDo8ffM2sFjdP5uC:-KMQQDo8ffM2sFjdP5uD:-679143926) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L243)):*
> ```Go
> 
> 	// If the soft/current limit is less than the hard limit, set the soft
> 	// limit to the hard/max limit.
> ```

We should set it to `min(rLimit.Max, suggestedOpenFileLimit)`.

---

*[storage/engine/rocksdb.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQQaYbj9mn4eygTtkl:-KMQQaYcihJAmz_GZPkB:-302428125) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/storage/engine/rocksdb.go#L61)):*
> ```Go
> 	// DefaultMaxOpenFiles is the default value for rocksDB's max_open_files
> 	// option.
> 	// TODO(bram): Once we upgrade to rocksdb 4.8, maybe change this to -1
> ```

Setting it to -1 will allow rocksdb to use the process' entire allocation of file descriptors. Why would we want to do that instead of setting a limit?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[server/context.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJCkAiu_c8xEbcS8z:-KMQbYkuq_1X5My5V5LB:561764072) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L229)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The `hard/max` terminology is slightly confusing. Can we just use `max`?
</details>
I like `max`, too. Also, the `current`/`max` pairing seems more descriptive than `soft`/`hard`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 7 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[server/context.go, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQIR7HFUPoNl6E6uCo:-KMQVyt3BsoScJHk0kuk:1379164192) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L62)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Should this be `= engine.DefaultMaxOpenFiles`?
> 
> `s/suggested/desired/g`?
</details>
Ah, good point.  I'll just use the engine value.

---

*[server/context.go, line 217 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQPQk7BotYErzCm18N:-KMQdHWgyO5yYL_qPR-0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L217)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We also need to allow room for network file descriptors.
</details>
Done.

---

*[server/context.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJCkAiu_c8xEbcS8z:-KMQY_fRW0RdtTvpAu9q:-1128982120) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L229)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The `hard/max` terminology is slightly confusing. Can we just use `max`?
</details>
Sure.  Done.

---

*[server/context.go, line 238 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQMxjdxGsSSrcEyTbH:-KMQdOvBHy6HNPbktaYX:1308588418) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L238)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> File descriptors consume some system-wide resources in the kernel. I think that even if the system sets the hard limit to "unlimited" we still want to set some reasonable limit for ourselves so that there's still an upper bound on how many descriptors we can open.
</details>
Added the upper bound.

---

*[server/context.go, line 243 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQQDo8ffM2sFjdP5uC:-KMQ_6EQPt5AHA0KMkbU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L243)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We should set it to `min(rLimit.Max, suggestedOpenFileLimit)`.
</details>
Done.

---

*[server/context.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQMhoqJKr32tJMGR8c:-KMQbjaFeSc42Rhclbvc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L253)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The grammar in this sentence seems off ("does not set to").
</details>
Done.

---

*[server/context.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQN7Y6yByy-TSc-3FZ:-KMQd0u7o7ZL2k5weBKX:1080555941) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L287)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> See `util.Pluralize()`.
</details>
removed as it was not needed

---

*[server/context.go, line 320 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQNBTj-vrhz812etfY:-KMQd9Hd0g3JFHiUntzj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L320)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/onDisk/physical/g` to match terminology above.
</details>
Done.

---

*[storage/engine/rocksdb.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQQaYbj9mn4eygTtkl:-KMQdCn6EC0Bs0gywVgv:1697881620) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/storage/engine/rocksdb.go#L61)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Setting it to -1 will allow rocksdb to use the process' entire allocation of file descriptors. Why would we want to do that instead of setting a limit?
</details>
Good point, removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 7 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQmaJF2l8QHUkav5xo:-KMQmaJF2l8QHUkav5xp:-1397852449) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L252)):*
> ```Go
> 	if rLimit.Cur < newCurrent {
> 		if log.V(1) {
> 			log.Infof("setting the current limit for open file descriptors (%d) to %d as it greatly increase performance",
> ```

s/increase/increases/ (or just remove this clause; it seems out of place for this message. In many cases increasing the rlimit is needed to meet the minimums, not just to improve performance)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-:-KMQnLGISjVq7-rflNEA:1417088721)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJCkAiu_c8xEbcS8z:-KMQdFZlYwS7HS814H-w:2133839414) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L229)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Sure.  Done.
</details>
Suggested revision: `max open file limit (%d) is under the minimum required (%d) \n%s`

---

*[server/context.go, line 223 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpAV0oIPPeuF3zwEl:-KMQpAV1NynXhLJN6fOS:223511801) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L223)):*
> ```Go
> 	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rLimit); err != nil {
> 		if log.V(1) {
> 			log.Infof("could not get rlimit, setting maxOpenFiles to the default value %d - %s", engine.DefaultMaxOpenFiles, err)
> ```

Not sure of our error message conventions, but grammatically, I'd use a semicolon instead of a comma here.

---

*[server/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQmaJF2l8QHUkav5xo:-KMQnfmRjW5qSJ6bA_Jb:2140260774) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L252)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/increase/increases/ (or just remove this clause; it seems out of place for this message. In many cases increasing the rlimit is needed to meet the minimums, not just to improve performance)
</details>
Suggested revision (piggybacking on Ben): `setting the current open file limit (%d) to %d`

---

*[server/context.go, line 265 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlM152h1xWyIK0yzd:-KMQlM1678ijZaifREKD:897132665) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L265)):*
> ```Go
> 		}
> 		if log.V(1) {
> 			log.Infof("the current limit for open file descriptors is now set to (%d)",
> ```

Similar to above. Suggested revision: `current open file limit is now %d`

---

*[server/context.go, line 272 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlbM2jqbEH9Q_ZmJH:-KMQlbM3mVuumXGgUt6z:-56105692) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L272)):*
> ```Go
> 	// The current open file descriptor limit is still too low.
> 	if rLimit.Cur < minimumOpenFileLimit {
> 		return 0, fmt.Errorf("current limit for open file descriptors is too low (%d) and cannot be increased, minimum required %d\n%s",
> ```

Suggested revision: `current open file limit (%d) is under the minimum required (%d) and cannot be increased \n%s`

---

*[server/context.go, line 285 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlzzUa3Elk6_WDFZ5:-KMQlzzUa3Elk6_WDFZ6:-197504117) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L285)):*
> ```Go
> 	// We're still below the recommended amount, we should always show a
> 	// warning.
> 	log.Warningf("the open file descriptors limit of %d is lower than the recommended amount of %d, this may decrease performance\n%s",
> ```

Revise to: `current open file limit (%d) is under the recommended limit (%d); this may decrease performance`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 65 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpRDmg-zkhuTj-2Ln:-KMQpRDmg-zkhuTj-2Lo:291409164) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L65)):*
> ```Go
> 	recommendedNetworkFileDescriptors = 5000
> 
> 	productionSettingsWebpage = "please see http://www.cockroachlabs.com/docs/recommended-production-settings.html for more details"
> ```

Didn't catch this before: We should change to `https`.

---

*[server/context.go, line 214 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpuqTYCyIj_7jE7tn:-KMQpuqTYCyIj_7jE7to:-290263293) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L214)):*
> ```Go
> 
> // setOpenFileLimit sets the current limit for open file descriptors to the max
> // limit if needed. Returns an error if the hard limit it too low. Returns the
> ```

Change "it too low" to "is too low".

---

*[server/context.go, line 236 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQqW9g9D5zsm31NIwz:-KMQqW9g9D5zsm31NIx-:517032511) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L236)):*
> ```Go
> 	}
> 
> 	// The current open file descriptor limit is higher than the recommended
> ```

Sentence would flow better if it started: "If the current open ..."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 6 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[server/context.go, line 229 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQJCkAiu_c8xEbcS8z:-KMQqCFaR3z2POAkTUcx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4327d9606a741a62c24ef7d2c91f84bf6075ee6/server/context.go#L229)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Suggested revision: `max open file limit (%d) is under the minimum required (%d) \n%s`
</details>
Done.

---

*[server/context.go, line 65 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpRDmg-zkhuTj-2Ln:-KMQqo8t7uB8fCq93S-M:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L65)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Didn't catch this before: We should change to `https`.
</details>
Done.

---

*[server/context.go, line 214 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpuqTYCyIj_7jE7tn:-KMQqtHZeGYr8WX3NWNA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L214)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Change "it too low" to "is too low".
</details>
Done.

---

*[server/context.go, line 223 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQpAV0oIPPeuF3zwEl:-KMQq1AXG_TBh0Hjj3w9:-819900810) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L223)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Not sure of our error message conventions, but grammatically, I'd use a semicolon instead of a comma here.
</details>
Big fan of the semicolon!  Done.

---

*[server/context.go, line 236 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQqW9g9D5zsm31NIwz:-KMQqzyLUZj4fqmFe29c:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L236)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Sentence would flow better if it started: "If the current open ..."
</details>
Done.

---

*[server/context.go, line 252 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQmaJF2l8QHUkav5xo:-KMQpKDPiOja60wXVxTR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L252)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/increase/increases/ (or just remove this clause; it seems out of place for this message. In many cases increasing the rlimit is needed to meet the minimums, not just to improve performance)
</details>
Done.

---

*[server/context.go, line 265 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlM152h1xWyIK0yzd:-KMQqQdjR40AQNGKBS0W:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L265)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Similar to above. Suggested revision: `current open file limit is now %d`
</details>
Done.

---

*[server/context.go, line 272 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlbM2jqbEH9Q_ZmJH:-KMQpgn-iPyZgiHNZb8P:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L272)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Suggested revision: `current open file limit (%d) is under the minimum required (%d) and cannot be increased \n%s`
</details>
Done.

---

*[server/context.go, line 285 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747#-KMQlzzUa3Elk6_WDFZ5:-KMQpx8QaLHiEwqAIEy5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c2ffbea30cebafa241a2bab11c62d2817d3636/server/context.go#L285)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Revise to: `current open file limit (%d) is under the recommended limit (%d); this may decrease performance`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7747)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Basic docs are here: https://www.cockroachlabs.com/docs/recommended-production-settings.html

Expanded instructions for increasing the file descriptors limit coming soon.
